src/interface/swarm.cpp

```diff
@@ -317,28 +354,537 @@ void Swarm::Defrag() {
-    pmb->par_for(
-        "Swarm::DefragInt", 0, max_active_index_, KOKKOS_LAMBDA(const int n) {
-          if (from_to_indices(n) >= 0) {
-            vec(from_to_indices(n)) = vec(n);
+  SwarmVariablePack<Real> vreal;
+  SwarmVariablePack<int> vint;
+  PackIndexMap rmap;
+  PackIndexMap imap;
+  vreal = PackAllVariablesReal(rmap);
+  vint = PackAllVariablesInt(imap);
+  int real_vars_size = realVector_.size();
+  int int_vars_size = intVector_.size();
+
+  pmb->par_for(
+      "Swarm::DefragVariables", 0, max_active_index_, KOKKOS_LAMBDA(const int n) {
+        if (from_to_indices(n) >= 0) {
+          for (int i = 0; i < real_vars_size; i++) {
+            vreal(i, from_to_indices(n)) = vreal(i, n);
           }
-        });
+          for (int i = 0; i < int_vars_size; i++) {
+            vint(i, from_to_indices(n)) = vint(i, n);
+          }
+        }
+      });
+
+  // Update max_active_index_
+  max_active_index_ = num_active_ - 1;
+}
+
+void Swarm::SetupPersistentMPI() {
+  vbswarm->SetupPersistentMPI();
+
+  // Index into neighbor blocks
+  auto pmb = GetBlockPointer();
+  auto neighborIndices_h = neighborIndices_.GetHostMirror();
+
+  // TODO(BRR) Checks against some current limitations
+  const int ndim = pmb->pmy_mesh->ndim;
+  auto mesh_bcs = pmb->pmy_mesh->mesh_bcs;
+  for (int n = 0; n < 2 * ndim; n++) {
+    PARTHENON_REQUIRE(mesh_bcs[n] == BoundaryFlag::periodic,
+                      "Only periodic boundaries supported right now!");
+  }
+
+  // Indicate which neighbor regions correspond to this meshblock
+  int kmin = 1;
+  int kmax = 3;
+  int jmin = 1;
+  int jmax = 3;
+  int imin = 1;
+  int imax = 3;
+  if (ndim < 3) {
+    kmin = 0;
+    kmax = 4;
+    if (ndim < 2) {
+      jmin = 0;
+      jmax = 4;
+    }
+  }
+  for (int k = kmin; k < kmax; k++) {
+    for (int j = jmin; j < jmax; j++) {
+      for (int i = imin; i < imax; i++) {
+        neighborIndices_h(k, j, i) = this_block_;
+      }
+    }
+  }
+
+  for (int n = 0; n < pmb->pbval->nneighbor; n++) {
+    NeighborBlock &nb = pmb->pbval->neighbor[n];
+
+    const int i = nb.ni.ox1;
+    const int j = nb.ni.ox2;
+    const int k = nb.ni.ox3;
+
+    if (ndim == 1) {
+      if (i == -1) {
+        neighborIndices_h(0, 0, 0) = n;
+      } else if (i == 0) {
+        neighborIndices_h(0, 0, 1) = n;
+        neighborIndices_h(0, 0, 2) = n;
+      } else {
+        neighborIndices_h(0, 0, 3) = n;
+      }
+    } else if (ndim == 2) {
+      if (i == -1) {
+        if (j == -1) {
+          neighborIndices_h(0, 0, 0) = n;
+        } else if (j == 0) {
+          neighborIndices_h(0, 1, 0) = n;
+          neighborIndices_h(0, 2, 0) = n;
+        } else if (j == 1) {
+          neighborIndices_h(0, 3, 0) = n;
+        }
+      } else if (i == 0) {
+        if (j == -1) {
+          neighborIndices_h(0, 0, 1) = n;
+          neighborIndices_h(0, 0, 2) = n;
+        } else if (j == 1) {
+          neighborIndices_h(0, 3, 1) = n;
+          neighborIndices_h(0, 3, 2) = n;
```

Comment:
This code would be a lot more readable if a lot of this logic was put into smaller subfunctions.

Reply:
How's the change I just pushed? Also, this logic will need to be re-done later for AMR (and I really want to think of a better way to do this for that case)

Reply:
It's an improvement.